### PR TITLE
Fix mul_hi in math_reference

### PR DESCRIPTION
**Pull request by aleksmesh**
Add check if we need to carry on overflow with signed types.

### CODE REVIEW

File: util/math_reference.cpp [hunk 1]
```diff
@@ -346,7 +346,16 @@ T mul_hi_signed(T a, T b) {
...
 
-  return (a >> msb) ^ (b >> msb) ? static_cast<T>(~hi) : hi;
+  T result = hi;
+  if ((a >> msb) ^ (b >> msb)) {
+    result = static_cast<T>(~hi);
+    // Check overflow to see if we need to carry
+    if (0 == static_cast<T>(a * b)) {
+      result += 1;
+    }
+  }
```
[the remark this answers is not in the document]
Comment (aleksmesh) [reply util/math_reference.cpp]: Comments added.

And I was wrong - it's not overflow checking, just checking that `U(r*s + ((cross2+cross1)<<shft)) == 0` (i.e. low half equal to zero) and we need to carry.
And I replaced it by `static_cast<T>(a*b) == 0` just for code simplification.

File: util/math_reference.cpp [hunk 2]
```diff
@@ -310,6 +310,15 @@ T mul_hi_unsigned(T x, T y) {
   return hi;
 }
 
+/**
+ * @brief Function to get high sizeof(T)*8 bits of the product of two signed T
```
Comment (keryell) [inline util/math_reference.cpp:314]: I guess the fact that `T` is an integer is missing. Can `T` be signed or unsigned?

Comment (keryell) [reply util/math_reference.cpp:314]: Add some `static_assert(std::is_integral_v...)`?

File: util/math_reference.cpp [hunk 3]
```diff
@@ -346,7 +356,17 @@ T mul_hi_signed(T a, T b) {
   lo >>= shft;
   hi += lo + (cross1 >> shft) + (cross2 >> shft);
 
-  return (a >> msb) ^ (b >> msb) ? static_cast<T>(~hi) : hi;
+  T result = hi;
+  // if result is negative
+  if ((a >> msb) ^ (b >> msb)) {
```
Comment (keryell) [inline util/math_reference.cpp:361]: All this obfuscated code does not make sense.
```suggestion
  if ((a < 0) ^ (b < 0)) {
```
Feel free to simplify all the code above too. :-)

File: util/math_reference.cpp [hunk 4]
```diff
@@ -336,6 +345,7 @@ T mul_hi_signed(T a, T b) {
   r = static_cast<U>(a_pos) & mask;
   s = static_cast<U>(b_pos) & mask;
 
+  // calc half products
```
Comment (keryell) [inline util/math_reference.cpp:348]: Comments are expected to be read by human beings
```suggestion
  // Compute half products
```

File: util/math_reference.cpp [hunk 5]
```diff
@@ -336,6 +345,7 @@ T mul_hi_signed(T a, T b) {
...
   lo = r * s;
   hi = p * q;
   cross1 = (p * s);
```
Comment (keryell) [inline util/math_reference.cpp:351]: You can remove the parenthesis while you are here. :-)

Comment (keryell) [reply util/math_reference.cpp:351]: Why not using `std::abs` above, replacing `typedef` by `using`, make `a_pos` and `b_pos` just a `U` and remove the `static_cast` later...

File: util/math_reference.cpp [hunk 6]
```diff
@@ -346,7 +356,17 @@ T mul_hi_signed(T a, T b) {
...
   hi += lo + (cross1 >> shft) + (cross2 >> shft);
 
-  return (a >> msb) ^ (b >> msb) ? static_cast<T>(~hi) : hi;
+  T result = hi;
+  // if result is negative
+  if ((a >> msb) ^ (b >> msb)) {
+    result = static_cast<T>(~hi);
+    // check that the low half is zero to see if we need to carry
+    if (0 == static_cast<T>(a * b)) {
+      result += 1;
```
Comment (keryell) [inline util/math_reference.cpp:365]: I need more time to understand this...

File: util/math_reference.cpp (hunk 2 above)
Comment (aleksmesh) [reply util/math_reference.cpp:314]: `mul_hi_signed` is only available inside math_reference.cpp and is only called below in `mul_hi` with signed input parameters, so it seems to me that we can do without static type checking.

File: util/math_reference.cpp (hunk 4 above)
Comment (aleksmesh) [reply util/math_reference.cpp]: Fixed. Thanks.

File: util/math_reference.cpp (hunk 5 above)
Comment (aleksmesh) [reply util/math_reference.cpp]: Fixed it. It certainly looks better. Thanks!

File: util/math_reference.cpp (hunk 3 above)
Comment (aleksmesh) [reply util/math_reference.cpp]: Done.

File: util/math_reference.cpp [hunk 7]
```diff
@@ -310,11 +310,20 @@ T mul_hi_unsigned(T x, T y) {
...
 }
 
+/**
+ * @brief Function to get high sizeof(T)*8 bits of the product of two signed T
+ *
+ * @tparam T signed type T
+ * @param a The first number of multiply
+ * @param b The second number of multiply
+ * @return T High sizeof(T)*8 bits of the sizeof(T*2)*8 bits result of the
+ * multiplication
```
Comment (keryell) [inline util/math_reference.cpp:320]:
```suggestion

   @tparam T signed type of operand
   @param a The first operand of multiply
   @param b The second operand of multiply
   @return T with high sizeof(T)*8 bits of the sizeof(T)*2*8 bits result of the
   multiplication
```

File: util/math_reference.cpp (hunk 6 above)
Comment (keryell) [reply util/math_reference.cpp:363]: OK, if the result of the multiplication of 2 non null operands is 0 there is a carry to propagate. If only one operand is 0, result is -1 and then adding 1 set it back to 0.
This looks like it works. :-)

**Review by keryell — APPROVED** [review]
Thanks. Just the comments to clarify.

File: util/math_reference.cpp (hunk 7 above)
Comment (aleksmesh) [reply util/math_reference.cpp]: Done. Thanks!

**Pull request merged**